app/models/orders_package.rb

```diff
@@ -20,4 +22,95 @@ def set_initial_state
       transition :requested => :designated
     end
   end
+
+  def self.update_partially_designated_item(package)
+    orders_package = OrdersPackage.find(package[:orders_package_id].to_i)
+    total_quantity = orders_package.quantity + package[:quantity].to_i
+    if(orders_package.state == "cancelled")
+      orders_package.update(quantity: total_quantity, state: "designated")
+    else
+      orders_package.update(quantity: total_quantity)
+    end
+  end
+
+  def self.undesignate_partially_designated_item(packages)
+    packages.each do |package|
+      quantity_to_reduce = package.last[:quantity].to_i
+      orders_package = OrdersPackage.find(package.last[:orders_package_id].to_i)
+      total_quantity = orders_package.quantity - quantity_to_reduce
+      if total_quantity == 0
+        orders_package.update(quantity: total_quantity, state: "cancelled")
+      else
+        orders_package.update(quantity: total_quantity, state: "designated")
+      end
+    end
+  end
+
+  def self.add_partially_designated_item(package)
+    self.create(
+      order_id: package[:order_id].to_i,
+      package_id: package[:package_id].to_i,
+      quantity: package[:quantity].to_i,
+      updated_by: User.current_user,
+      state: "designated"
+      )
+  end
+
+  def self.find_packages(order_id, package_id)
+    orders_packages = OrdersPackage.where("order_id = ? and package_id = ?", order_id, package_id)
+  end
+
+  def self.find_records(order_id)
+    OrdersPackage.where(order_id: order_id)
+  end
+
+  def self.get_total_dispatched_qty(package_id)
+    packages = OrdersPackage.where("package_id = ? and state = ?", package_id, "dispatched")
+    total_dispatched_qty = 0
+    packages.each do |orders_package|
+      total_dispatched_qty += orders_package.quantity
+    end
+    total_dispatched_qty
+  end
+
+  def self.update_designation(orders_package, order_id)
```

Comment:
I wonder if these `self` methods can be refactored into instance methods. In fact, for some, you can iterate over objects and use existing active_record update_columns.

So instead of

    OrdersPackage.update_designation(orders_package, order_id)

Can have something more like

    class OrdersPackage
      def update_designation(order_id)
        update(order_id, order_id)
      end
    end

    orders_package.update_designation(order_id)

In general, I think class methods need to be avoided where possible.